rdmd.d

```diff
@@ -916,3 +882,74 @@ void yap(size_t line = __LINE__, T...)(auto ref T stuff)
     debug stderr.writeln(line, ": ", stuff);
     else stderr.writeln(stuff);
 }
+
+/**
+Used to wrap filesystem operations that should also be logged with the
+yap function. These functions allow the the filenane to be given once
+so the log statement always matches the operation. Using it also guarantees
+you won't forget to include a `yap` alongside any file operation you want
+to be logged.
+*/
+static struct Yap
+{
+    static import std.file;
+
+    static auto opDispatch(string func, T...)(T args)
+    {
+        static if (func.endsWith("IfLive"))
+        {
+            enum fileFunc = func[0 .. $ - "IfLive".length];
+            enum skipOnDryRun = true;
+        }
+        else
+        {
+            enum fileFunc = func;
+            enum skipOnDryRun = false;
+        }
+
+        static if ([fileFunc].findAmong(["exists", "timeLastModified", "isFile"]))
+            yap("stat ", args[0]);
+        else static if (fileFunc == "rename")
+            yap("mv ", args[0], " ", args[1]);
+        else static if ([fileFunc].findAmong(["remove", "mkdirRecurse", "rmdirRecurse", "dirEntries", "write"]))
+            yap(fileFunc, " ", args[0]);
+        else static assert(0, "Yap.opDispath has not implemented " ~ fileFunc);
+
+        static if (skipOnDryRun)
+        {
+            if (dryRun)
+                return;
+        }
+        mixin("return std.file." ~ fileFunc ~ "(args);");
+    }
+    // Special case, combine these 2 operations so you don't end up with
+    // multiple yap statements in the log for the same file.
+    static bool existsAsFile(const(char)[] name)
+    {
+        yap("stat ", name);
+        return std.file.exists(name) && std.file.isFile(name);
+    }
+    static void touchIfLive(const(char)[] name)
+    {
+        yap("touch ", name);
+        if (!dryRun)
+            std.file.write(name, "");
+    }
+    static void removeOrRenameIfLive(const(char)[] name)
+    {
+        yap("remove ", name);
+        if (!dryRun)
+        {
+            try
+                std.file.remove(name);
+            catch (FileException e)
+            {
+                // This can occur on Windows if the executable is locked.
+                // Although we can't delete the file, we can still rename it.
+                auto target = "%s.%s-%s.old".format(name,
+                    Clock.currTime.stdTime, thisProcessID);
+                Yap.rename(name, target);
+            }
+        }
+    }
```

Comment:
This is completely not what I meant :(

Why not keep the `try/catch` in the logic code above, and use `removeIfLive` here?

Reply:
ah, not a bad idea, will do